### PR TITLE
Refactor of `changeSensorConfig()` and additional improvements

### PR DESCRIPTION
**Pull request by SwoopX**
Primarily, this change is intended to clean up the code chaos in `changeSensorConfig()` and streamline API behavior upon sensor config changes.

- Use `QLatin1String()` wherever possible throughout `rest_sensors.cpp`
- Always respond with HTTP code 200 when a request has been received by the API
- Now respond with errors homogeneously in the response body when:
  - The defined data type for the resource item is incorrect (e.g. a number when true/false is expected)
  - The resource item is not modifiable (e.g. battery)
  - An attribute could not be set (meaning the request couldn't be queued by deCONZ)
  - A resource item is not available for the sensor (e.g. `displayflipped` for a thermostat)
- Introduce a mandatory type check for resource items, so they do not accept anything else than the defined type (e.g. `1` to express `true` is now disallowed)
- Introduce `std::array`s for allowed string values for certain resource items
- Introduce a fast search function to return a found key value (if any) for the valid string values in the defined arrays
- Move several constants to seperate header files for reuse
- Eliminate duplicates and inconsistencies
- Set valid range for `RConfigMelody` (NAS-AB02B0 Siren)
- Use `RConfigTempMinThreshold`, `RConfigTempMaxThreshold`, `RConfigHumiMinThreshold` and `RConfigHumiMaxThreshold` for NAS-AB02B0 Siren (instead of just `RConfigTempThreshold` and `RConfigHumiThreshold`)
- Fix setting `RConfigMode` for SEA801-ZIGBEE TRV

This PR will majorly affect thermostat devices, as the vast majority of changeable resource items is related to those devices. At the time of writing this, the changes have been tested with:
- Danfoss Ally (thermostat)
- Eurotronic Spirit (thermostat)
- Hue motion sensor

Some Tuya thermostats will temporarily not have the possibility to set `windowopen_set` in case a comma separated string was previously used to set this capability (e.g. `true, 10, 11`). However, this can be added subsequently where required.

### CODE REVIEW

**Review by manup — COMMENTED** [review]
Overall it looks good, added a few comments

File: thermostat.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+// Thermostat cluster, Control Sequence of Operation (0x001B)
+#define COOLING_ONLY                            0x00
+#define COOLING_WITH_REHEAT                     0x01
+#define HEATING_ONLY                            0x02
+#define HEATING_WITH_REHEAT                     0x03
+#define COOLING_AND_HEATING_4PIPES              0x04
+#define COOLING_AND_HEATING_4PIPES_WITH_REHEAT  0x05
+
+static const std::array<KeyValMap, 6> RConfigModeLegrandValues = { { {QLatin1String("confort"), 0}, {QLatin1String("confort-1"), 1}, {QLatin1String("confort-2"), 2},
+                                                              {QLatin1String("eco"), 3}, {QLatin1String("hors gel"), 4}, {QLatin1String("off"), 5} } };
```
[the remark this answers is not in the document]
Comment (SwoopX) [reply thermostat.h]: Because it's French 🤷‍♂️ 

Done.

File: thermostat.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+#define CONTROL_SEQUENCE_OF_OPERATION           0x001B
+#define SYSTEM_MODE                             0x001C
+#define THERMOSTAT_PROGRAMMING_OPERATION_MODE   0x0025
+#define AC_LOUVER_POSITION                      0x0045
+#define TEMPERATURE_MEASUREMENT                 0x0403 // ELKO specific
+#define DEVICE_ON                               0x0406 // ELKO specific
+#define CHILD_LOCK                              0x0413 // ELKO specific
+#define CURRENT_TEMPERATURE_SETPOINT            0x4003 // Eurotronic specific
+#define EXTERNAL_OPEN_WINDOW_DETECTED           0x4003 // Danfoss specific
+#define HOST_FLAGS                              0x4008 // Eurotronic specific
```
[the remark this answers is not in the document]
Comment (SwoopX) [reply thermostat.h]: Done.

File: simple_metering.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+#define NORWEGIAN_HAN_EXTRA_LOAD        0x0201
+#define AIDON_METER                     0x0202
+#define KAIFA_KAMSTRUP_METERS           0x0203
+#define AUTO_DETECT                     0x0204
+
+static const std::array<KeyValMapInt, 8> RConfigInterfaceModeValuesZHEMI = { { {1, PULSE_COUNTING_ELECTRICITY}, {2, PULSE_COUNTING_GAS}, {3, PULSE_COUNTING_WATER},
+                                                                               {4, KAMSTRUP_KMP}, {5, LINKY}, {6, DLMS_COSEM}, {7, DSMR_23}, {8, DSMR_40} } };
+
+static const std::array<KeyValMapInt, 5> RConfigInterfaceModeValuesEMIZB = { { {1, NORWEGIAN_HAN}, {2, NORWEGIAN_HAN_EXTRA_LOAD}, {3, AIDON_METER},
+                                                                               {4, KAIFA_KAMSTRUP_METERS}, {5, AUTO_DETECT} } };
```
[the remark this answers is not in the document]
Comment (SwoopX) [reply simple_metering.h]: Done.

File: rest_sensors.cpp [hunk 4]
```diff
@@ -716,16 +721,19 @@ int DeRestPluginPrivate::changeSensorConfig(const ApiRequest &req, ApiResponse &
...
-            if (!isClip && (rid.suffix == RConfigBattery || rid.suffix == RConfigReachable))
-            {
-                // changing battery or reachable of zigbee sensors is not allowed, trigger error
-            }
-            else if (rid.suffix == RConfigPending || rid.suffix == RConfigSensitivityMax || rid.suffix == RConfigHostFlags)
+            updated = false;
+
+            // Changing these values of zigbee sensors is not allowed, read-only.
+            if (rid.suffix == RConfigPending || rid.suffix == RConfigSensitivityMax || rid.suffix == RConfigHostFlags || rid.suffix == RConfigLastChangeAmount ||
+                rid.suffix == RConfigLastChangeSource || rid.suffix == RConfigLastChangeTime || rid.suffix == RConfigEnrolled || rid.suffix == RConfigOn ||
```
[the remark this answers is not in the document]
Comment (SwoopX) [reply rest_sensors.cpp]: Good point, done.

File: rest_sensors.cpp [hunk 5]
```diff
@@ -740,561 +748,261 @@ int DeRestPluginPrivate::changeSensorConfig(const ApiRequest &req, ApiResponse &
...
-                            if (item->lastChanged() == item->lastSet())
-                            {
-                                updated = true;
-                            }
-                        }
-                    }
-                    else
+                if (rid.suffix == RConfigDeviceMode) // String
+                {
+                    if (RConfigDeviceModeValues.indexOf(val.toString()) >= 0)
```
[the remark this answers is not in the document]
Comment (SwoopX) [reply rest_sensors.cpp]: Indeed, done.

File: rest_sensors.cpp [hunk 6]
```diff
@@ -533,46 +537,46 @@ int DeRestPluginPrivate::updateSensor(const ApiRequest &req, ApiResponse &rsp)
...
-        if (!((pi.key() == "name") || (pi.key() == "modelid") || (pi.key() == "swversion")
-             || (pi.key() == "type")  || (pi.key() == "uniqueid")  || (pi.key() == "manufacturername")
-             || (pi.key() == "state")  || (pi.key() == "config")
-             || (pi.key() == "mode" && (sensor->modelId() == "Lighting Switch" || sensor->modelId().startsWith(QLatin1String("SYMFONISK"))))))
+        if (!((pi.key() == QLatin1String("name")) || (pi.key() == QLatin1String("modelid")) || (pi.key() == QLatin1String("swversion"))
+             || (pi.key() == QLatin1String("type"))  || (pi.key() == QLatin1String("uniqueid"))  || (pi.key() == QLatin1String("manufacturername"))
+             || (pi.key() == QLatin1String("state"))  || (pi.key() == QLatin1String("config"))
+             || (pi.key() == QLatin1String("mode") && (sensor->modelId() == QLatin1String("Lighting Switch") || sensor->modelId().startsWith(QLatin1String("SYMFONISK"))))))
         {
             rsp.list.append(errorToMap(ERR_PARAMETER_NOT_AVAILABLE, QString("/sensors/%2").arg(pi.key()), QString("parameter, %1, not available").arg(pi.key())));
```
[the remark this answers is not in the document]
Comment (SwoopX) [reply rest_sensors.cpp]: Done.

File: rest_sensors.cpp [hunk 7]
```diff
@@ -288,15 +292,15 @@ int DeRestPluginPrivate::createSensor(const ApiRequest &req, ApiResponse &rsp)
...
-        rsp.list.append(errorToMap(ERR_SENSOR_LIST_FULL , QString("/sensors/"), QString("The Sensor List has reached its maximum capacity of %1 sensors").arg(MAX_SENSORS)));
+        rsp.list.append(errorToMap(ERR_SENSOR_LIST_FULL, QLatin1String("/sensors/"), QString("The Sensor List has reached its maximum capacity of %1 sensors").arg(MAX_SENSORS)));
         rsp.httpStatus = HttpStatusBadRequest;
         return REQ_READY_SEND;
     }
 
     //check required parameter
-    if ((!(map.contains("name")) || !(map.contains("modelid")) || !(map.contains("swversion")) || !(map.contains("type")) || !(map.contains("uniqueid")) || !(map.contains("manufacturername"))))
+    if ((!(map.contains(QLatin1String("name"))) || !(map.contains(QLatin1String("modelid"))) || !(map.contains(QLatin1String("swversion"))) || !(map.contains(QLatin1String("type"))) || !(map.contains(QLatin1String("uniqueid"))) || !(map.contains(QLatin1String("manufacturername")))))
     {
```
[the remark this answers is not in the document]
Comment (SwoopX) [reply rest_sensors.cpp]: Done.

File: thermostat.h [hunk 8]
```diff
@@ -0,0 +1,62 @@
+#ifndef THERMOSTAT_H
+#define THERMOSTAT_H
+
+#include <QString>
+#include <array>
```
[the remark this answers is not in the document]
Comment (SwoopX) [reply thermostat.h]: Ok, done.

**Pull request merged**